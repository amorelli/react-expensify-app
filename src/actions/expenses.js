import uuid from 'uuid';
import database from '../firebase/firebase';

// Changes the Redux store
export const addExpense = (expense = {}) => ({
    type: 'ADD_EXPENSE',
    expense
  });

// Asynchronous function, fetches data from Firebase
export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '', 
      note = '', 
      amount = 0, 
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id 
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};

export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    });
  };
};

// Setup expense action object with data
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

// Gets initial expense data from Firebase
export const startSetExpenses = () => {
  return (dispatch, getState) => {
    // Fetch data
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
      // Parse data
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      // Get data in Redux
      dispatch(setExpenses(expenses));
    });
  };
};