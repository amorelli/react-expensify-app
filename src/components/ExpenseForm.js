import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      focused: false,
      error: '',
    };
  }
  // Description and Note are tracked by the state and updated
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }
  onAmountChange = (e) => {
    const amount = e.target.value;
    if ( !amount || amount.match(/^\d{1,}(\.\d{0,2})?$/) ) { // Matches a string starting (^) with a decimal (\d) continue to match decimals (*)
      this.setState(() => ({ amount }));       // Matches 0 to 1 groups ( ()? ) of a period (\.) and between 0 and 2 decimals (\d{0,2}) 
    }                                          // then stop matching ($)
  };
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({focused}) => {
    this.setState(() => ({ focused }));
  };
  onSubmit = (e) => {
    e.preventDefault(); // Prevent page refresh, handle form submission with JS
    // Error message if Description & Amount are blank
    if ( !this.state.description || !this.state.amount ) {
      this.setState(() => ({ error: 'Please provide Description and Amount.' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({ // This object is passed up to AddExpensePage when AddExpensePage calls onSubmit
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error && <div>{this.state.error}</div>}
        <form onSubmit={this.onSubmit}>
          <input 
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description} // Grab the description value from the state
            onChange={this.onDescriptionChange} // Set the state of the description value
          />
          <input 
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker 
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.focused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense."  
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}