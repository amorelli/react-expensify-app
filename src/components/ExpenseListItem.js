import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt }) => (
  <div>
    <h3>
      <Link to={`/edit/${id}`}>
        {description}
      </Link>
    </h3>
    <p>
      {numeral(amount / 100).format('$0,0.00')} - {moment(createdAt).format('MMMM Do, YYYY')}
    </p>
  </div>
);

export default ExpenseListItem;
