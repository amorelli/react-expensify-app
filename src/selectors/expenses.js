import moment from 'moment';

// Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => { // Filter creates an array with all elements that pass the conditions provided by the function
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase()); // Text filter matches if text is in description

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    }
    if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};