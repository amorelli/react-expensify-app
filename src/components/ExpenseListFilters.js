import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filters';

// We read off the store from this component
export class ExpenseListFilters extends React.Component {

  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  handleFocus = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value); // Set the store text filter based on the user-input value
  };

  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount(); // As we change the select, the store value changes, as the store changes, the value of select changes
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input 
              type="text"
              className="text-input"
              placeholder="Search Expenses"
              value={this.props.filters.text} // Value set based on the text filter from the store
              onChange={this.onTextChange} />
          </div>
          <div className="input-group__item">
            <select 
              className="select"
              value={this.props.filters.sortBy} 
              onChange={this.onSortChange}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker 
              startDate={this.props.filters.startDate}
              startDateId='start'
              endDateId='end'
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.handleFocus}
              showClearDates={true}
              numberOfMonths={1}
              isOutsideRange={() => false}
            />
          </div>
        </div>
      </div>
    );
  }
};

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchtoProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate))
});

export default connect(mapStateToProps, mapDispatchtoProps)(ExpenseListFilters);