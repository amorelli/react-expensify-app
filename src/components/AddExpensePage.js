import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses'; // Import startAddExpense action generator

export class AddExpensePage extends React.Component {
  
  onSubmit = (expense) => { // Expense object received from ExpenseForm {description, amount, note, createdAt}
    this.props.startAddExpense(expense);
    this.props.history.push('/'); // Programmatically change pages, takes target page argument
  };

  render() {
    return (
      <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Add Expense</h1>
        </div>
      </div>
        <div className="content-container">
          <ExpenseForm 
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage); // With connect we have access to props.dispatch