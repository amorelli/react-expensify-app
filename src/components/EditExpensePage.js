import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import Modal from 'react-modal';
import moment from 'moment';

export class EditExpensePage extends React.Component {
 
  state = {
      selectedOption: false
  };

  onSubmit = (expense) => {
      this.props.startEditExpense(this.props.expense.id, expense);
      this.props.history.push('/');
  };

  openModal = () => {
      this.setState(() => ({selectedOption: true}));
  };

  closeModal = () => {
      this.setState(() => ({selectedOption: false}));
  };


  removeExpense = () => {
      this.props.startRemoveExpense({id: this.props.expense.id});
      this.props.history.push('/');
  };

  render() {
      return (
          <div>
              <div className="page-header">
                  <div className="content-container">
                      <h1 className="page-header__title">Edit Expense</h1>
                  </div>
              </div>
              
              <div className="content-container">
                  <ExpenseForm 
                      expense={this.props.expense}
                      onSubmit={this.onSubmit}
                  />
                  <button className="button button--secondary" onClick={this.openModal}>Remove Expense</button>
              </div>
              <Modal
                  isOpen={this.state.selectedOption}
                  contentLabel="Removing Expense"
                  onRequestClose={this.closeModal}
                  closeTimeoutMS={200}
                  className="modal"
              >
                      <p>Are you sure you want to remove <span>{this.props.expense.description}</span> from your expenses?</p>
                      <button className="button" onClick={this.removeExpense}>Yes</button>
                      <button className="button button--secondary" onClick={this.closeModal}>No</button>
              </Modal>
          </div>
      );
  }
}

const mapDispatchToProps = (dispatch, props) => (
  {
      startEditExpense: (expenseId, expense) => dispatch(startEditExpense(expenseId, expense)),
      startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
  }
);

const mapStateToProps = (state, props) => {
  return {
      expense: state.expenses.find((expense) => (expense.id === props.match.params.id))
  };
};  

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);

