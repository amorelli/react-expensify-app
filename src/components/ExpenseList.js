import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// Our component with props from the store
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      {
        props.expenses.length === 0 ? (
          <div className="list-item list-item--message">
            <span>No Expenses</span>
          </div>
        ) : (
          props.expenses.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense}/>
          })
        )
      }
    </div>
  </div>
);

// Define the things we want to get off of the store using react-redux connect
// This is a Higher Order Component
// Store information passed into component
// Store's state gets passed in
// As store changes, this is automatically re-run, getting new values to the component
// Connected components are reactive
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
