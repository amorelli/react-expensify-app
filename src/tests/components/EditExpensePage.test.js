import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  startRemoveExpense = jest.fn();
  startEditExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage 
      expense={expenses[1]} 
      startEditExpense={startEditExpense} 
      startRemoveExpense={startRemoveExpense} 
      history={history} 
    />
  );
});

// EditExpensePage Snapshot
it('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

// startEditExpense
it('should handle startEditExpense', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[1]);

  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startEditExpense).toHaveBeenLastCalledWith(expenses[1].id, expenses[1]);
});

// startRemoveExpense
it('should handle startRemoveExpense', () => {
  wrapper.find('button').at(1).prop('onClick')(expenses[1]);

  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(startRemoveExpense).toHaveBeenLastCalledWith({ id: expenses[1].id });
});