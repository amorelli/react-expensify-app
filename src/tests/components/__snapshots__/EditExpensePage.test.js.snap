// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 109500,
          "createdAt": -345600000,
          "description": "Rent",
          "id": "2",
          "note": "",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    className="modal"
    closeTimeoutMS={200}
    contentLabel="Removing Expense"
    isOpen={false}
    onRequestClose={[Function]}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    role="dialog"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
  >
    <p>
      Are you sure you want to remove 
      <span>
        Rent
      </span>
       from your expenses?
    </p>
    <button
      className="button"
      onClick={[Function]}
    >
      Yes
    </button>
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      No
    </button>
  </Modal>
</div>
`;
